File: problems/ruby/010721-remove-duplicates/code.rb
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next

  def initialize(val)
    @val = val
    @next = nil
  end
end

class SinglyLinkedList
  attr_accessor :head, :tail

  def initialize
    @head = nil
    @tail = nil
  end

  def push(val)
    if !@head
      @head = ListNode.new(val)
      @tail = @head
    else
      @tail.next = ListNode.new(val)
      @tail = @tail.next
    end
  end

  def print
    walker = @head
    while walker
      puts walker.val
      walker = walker.next
    end
  end
end

# https://leetcode.com/problems/remove-duplicates-from-sorted-list/
#
# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
  first_walker = head

  while first_walker
    second_walker = first_walker
    while second_walker.next && second_walker.val == second_walker.next.val
      second_walker = second_walker.next
    end
    first_walker.next = second_walker.next
    first_walker = first_walker.next
  end

  head
end

list = SinglyLinkedList.new

[1, 1, 1, 2, 2, 3, 3, 3, 4, 5, 5].each do |num|
  list.push(num)
end

delete_duplicates(list.head)

list.print

File: problems/javascript/01072021-remove-duplicates/code.js
/**
 * Singly linked list definition
 */
class ListNode {
  constructor(val) {
    this.val = val
    this.next = null
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null
    this.tail = null
  }

  push(val) {
    let newNode = new ListNode(val)

    if (!this.head) {
      this.head = newNode
      this.tail = this.head
    } else {
      this.tail.next = newNode
      this.tail = this.tail.next
    }
  }

  print() {
    let walker = this.head
    while (walker) {
      console.log(walker.val)
      walker = walker.next
    }
  }
}

/**
 * https://leetcode.com/problems/remove-duplicates-from-sorted-list/
 *
 * @param {ListNode} head
 * @returns {ListNode}
 */

const deleteDuplicates = (head) => {
  let firstWalker = head

  while (firstWalker) {
    let secondWalker = firstWalker
    while (secondWalker.val === secondWalker.next?.val) {
      secondWalker = secondWalker.next
    }
    firstWalker.next = secondWalker.next
    firstWalker = firstWalker.next
  }

  return head
}

const list = new SinglyLinkedList()

const values = [1, 1, 1, 2, 2, 3, 3, 3, 4, 5, 5]
values.forEach((num) => list.push(num))

deleteDuplicates(list.head)

list.print()
